pkg/controllers/provisioning/scheduling/topology_test.go

```diff
@@ -400,6 +400,29 @@ var _ = Describe("Topology", func() {
 			}
 			Expect(nodeNames).To(HaveLen(1))
 		})
+		It("should respect minDomains constraints", func() {
+			if env.Version.Minor() < 24 {
+				Skip("MinDomains TopologySpreadConstraint is only available starting in K8s >= 1.24.x")
+			}
+			var minDomains int32 = 4
+			provisioner.Spec.Requirements = []v1.NodeSelectorRequirement{
+				{Key: v1.LabelTopologyZone, Operator: v1.NodeSelectorOpIn, Values: []string{"test-zone-1", "test-zone-2", "test-zone-3"}}}
+			topology := []v1.TopologySpreadConstraint{{
+				TopologyKey:       v1.LabelTopologyZone,
+				WhenUnsatisfiable: v1.DoNotSchedule,
+				LabelSelector:     &metav1.LabelSelector{MatchLabels: labels},
+				MaxSkew:           1,
+				MinDomains:        &minDomains,
+			}}
+			ExpectApplied(ctx, env.Client, provisioner)
+			ExpectProvisioned(ctx, env.Client, cluster, cloudProvider, prov,
+				test.UnschedulablePod(test.PodOptions{ObjectMeta: metav1.ObjectMeta{Labels: labels}, TopologySpreadConstraints: topology}),
+				test.UnschedulablePod(test.PodOptions{ObjectMeta: metav1.ObjectMeta{Labels: labels}, TopologySpreadConstraints: topology}),
+				test.UnschedulablePod(test.PodOptions{ObjectMeta: metav1.ObjectMeta{Labels: labels}, TopologySpreadConstraints: topology}),
+				test.UnschedulablePod(test.PodOptions{ObjectMeta: metav1.ObjectMeta{Labels: labels}, TopologySpreadConstraints: topology}),
+			)
+			ExpectSkew(ctx, env.Client, "default", &topology[0]).To(ConsistOf(1, 1, 1))
```

Comment:
@jonathan-innis @tzneal I think this PR is almost ready to go

What's failing now is that this test scenario is passing in a podDomains data structure that isn't what I expected; I'm expecting the nodeSelectorRequirement above would render as 3 domains, which would not fulfill the minDomains requirement in this test scenario. Instead we're getting this:

`domains: &scheduling.Requirement{Key:"topology.kubernetes.io/zone", complement:true, values:sets.String{}, greaterThan:(*int)(nil), lessThan:(*int)(nil)}`

Which yields a value of zero domains.

I wonder if our `provisioner.Spec.Requirements` assignment in L408 above isn't passing through to our `ExpectProvisioned` flow? Or maybe there's a better way of declaring this requirement here?

Reply:
I just looked over the code and gave a quick review. From what I can grok, it looks like the logic is slightly off. From what I read, you are doing a `minDomains` comparison against the pod-supported domains (that's what `domains`) is in the `minDomainCount` function. What we should be doing is checking the number of domains that the pod supports (this should basically be the count of iterations where `domains.Has(domain)` succeeds and then validate if that number of supported pod domains across the cluster is less than the number of `minDomains`. If it is, then we should `return 0`

Reply:
That makes sense, I think that was my initial attempt [here](https://github.com/aws/karpenter-core/pull/330#pullrequestreview-1421455632) but I probably misunderstood your comment!

Reply:
We have a passing test scenario with that change, thanks! Thoughts on this net new "should respect minDomains constraints" case being sufficient for this new feature?

Reply:
> That makes sense, I think that was my initial attempt [here](https://github.com/aws/karpenter-core/pull/330#pullrequestreview-1421455632) but I probably misunderstood your comment!

It's more likely that I got confused without looking into the details really critically. Yes, you're right that the initial attempt was correct. I naively assumed that the code worked differently. My fault! 🎉

Reply:
I recall my initial attempt had too many spokes, we're in a good place now. :)

Reply:
> Thoughts on this net new "should respect minDomains constraints" case being sufficient for this new feature

The only other test cases I could think of are cases with the `Exists` operator or with custom domains that are not part of the well-known label set. 

I'd imagine that doing something like custom domains should behave the same as the test that you currently have setup, so there's probably a marginal test coverage improvement there, but it might be worth adding.

We should probably also validate that we succeed to schedule everything when the number of `minDomains` is less than the total possible domains, just to make sure we are covering all our bases and a small bug doesn't slip into the logic.

@tzneal Did you have any additional thoughts on test coverage?

Reply:
Done